projects/util-linux-2.40.2/sys-utils/choom.1
'\" t
.\"     Title: choom
.\"    Author: [see the "AUTHOR(S)" section]
.\" Generator: Asciidoctor 2.0.20
.\"      Date: 2024-04-24
.\"    Manual: User Commands
.\"    Source: util-linux 2.40.2
.\"  Language: English
.\"
.TH "CHOOM" "1" "2024-04-24" "util\-linux 2.40.2" "User Commands"
.ie \n(.g .ds Aq \(aq
.el       .ds Aq '
.ss \n[.ss] 0
.nh
.ad l
.de URL
\fI\\$2\fP <\\$1>\\$3
..
.als MTO URL
.if \n[.g] \{\
.  mso www.tmac
.  am URL
.    ad l
.  .
.  am MTO
.    ad l
.  .
.  LINKSTYLE blue R < >
.\}
.SH "NAME"
choom \- display and adjust OOM\-killer score.
.sp
\fBchoom\fP \fB\-p\fP \fIPID\fP
.sp
\fBchoom\fP \fB\-p\fP \fIPID\fP \fB\-n\fP \fInumber\fP
.sp
\fBchoom\fP \fB\-n\fP \fInumber\fP [\-\-] \fIcommand\fP [\fIargument\fP ...]
.SH "DESCRIPTION"
.sp
The \fBchoom\fP command displays and adjusts Out\-Of\-Memory killer score setting.
.SH "OPTIONS"
.sp
\fB\-p\fP, \fB\-\-pid\fP \fIpid\fP
.RS 4
Specifies process ID.
.RE
.sp
\fB\-n\fP, \fB\-\-adjust\fP \fIvalue\fP
.RS 4
Specify the adjust score value.
.RE
.sp
\fB\-h\fP, \fB\-\-help\fP
.RS 4
Display help text and exit.
.RE
.sp
\fB\-V\fP, \fB\-\-version\fP
.RS 4
Print version and exit.
.RE
.SH "NOTES"
.sp
Linux kernel uses the badness heuristic to select which process gets killed in out of memory conditions.
.sp
The badness heuristic assigns a value to each candidate task ranging from 0 (never kill) to 1000 (always kill) to determine which process is targeted. The units are roughly a proportion along that range of allowed memory the process may allocate from based on an estimation of its current memory and swap use. For example, if a task is using all allowed memory, its badness score will be 1000. If it is using half of its allowed memory, its score will be 500.
.sp
There is an additional factor included in the badness score: the current memory and swap usage is discounted by 3% for root processes.
.sp
The amount of "allowed" memory depends on the context in which the oom killer was called. If it is due to the memory assigned to the allocating task\(cqs cpuset being exhausted, the allowed memory represents the set of mems assigned to that cpuset. If it is due to a mempolicy\(cqs node(s) being exhausted, the allowed memory represents the set of mempolicy nodes. If it is due to a memory limit (or swap limit) being reached, the allowed memory is that configured limit. Finally, if it is due to the entire system being out of memory, the allowed memory represents all allocatable resources.
.sp
The adjust score value is added to the badness score before it is used to determine which task to kill. Acceptable values range from \-1000 to +1000. This allows userspace to polarize the preference for oom killing either by always preferring a certain task or completely disabling it. The lowest possible value, \-1000, is equivalent to disabling oom killing entirely for that task since it will always report a badness score of 0.
.sp
Setting an adjust score value of +500, for example, is roughly equivalent to allowing the remainder of tasks sharing the same system, cpuset, mempolicy, or memory controller resources to use at least 50% more memory. A value of \-500, on the other hand, would be roughly equivalent to discounting 50% of the task\(cqs allowed memory from being considered as scoring against the task.
.SH "AUTHORS"
.sp
.MTO "kzak\(atredhat.com" "Karel Zak" ""
.SH "SEE ALSO"
.sp
\fBproc\fP(5)
.SH "REPORTING BUGS"
.sp
For bug reports, use the issue tracker at \c
.URL "https://github.com/util\-linux/util\-linux/issues" "" "."
.SH "AVAILABILITY"
.sp
The \fBchoom\fP command is part of the util\-linux package which can be downloaded from \c
.URL "https://www.kernel.org/pub/linux/utils/util\-linux/" "Linux Kernel Archive" "."

projects/util-linux-2.40.2/sys-utils/choom.c
/*
 * choom - Change OOM score setting
 *
 * Copyright (C) 2018 Karel Zak <kzak@redhat.com>
 *
 * This program is free software; you can redistribute it and/or modify
 * it under the terms of the GNU General Public License as published by
 * the Free Software Foundation; either version 2 of the License, or
 * (at your option) any later version.
 *
 * This program is distributed in the hope that it would be useful,
 * but WITHOUT ANY WARRANTY; without even the implied warranty of
 * MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
 * GNU General Public License for more details.
 *
 * You should have received a copy of the GNU General Public License along
 * with this program; if not, write to the Free Software Foundation, Inc.,
 * 51 Franklin Street, Fifth Floor, Boston, MA 02110-1301 USA.
 */
#include <stdio.h>
#include <stdlib.h>
#include <string.h>
#include <unistd.h>
#include <getopt.h>
#include <errno.h>

#include "nls.h"
#include "c.h"
#include "path.h"
#include "strutils.h"
#include "closestream.h"

static void __attribute__((__noreturn__)) usage(void)
{
	FILE *out = stdout;
	fputs(USAGE_HEADER, out);
	fprintf(out,
	      _(" %1$s [options] -p pid\n"
		" %1$s [options] -n number -p pid\n"
		" %1$s [options] -n number [--] command [args...]]\n"),
		program_invocation_short_name);

	fputs(USAGE_SEPARATOR, out);
	fputs(_("Display and adjust OOM-killer score.\n"), out);

	fputs(USAGE_OPTIONS, out);
	fputs(_(" -n, --adjust <num>     specify the adjust score value\n"), out);
	fputs(_(" -p, --pid <num>        process ID\n"), out);
	fputs(USAGE_SEPARATOR, out);
	fprintf(out, USAGE_HELP_OPTIONS(24));
	fprintf(out, USAGE_MAN_TAIL("choom(1)"));
	exit(EXIT_SUCCESS);
}

static int get_score(struct path_cxt *pc)
{
	int ret;

	if (ul_path_read_s32(pc, &ret, "oom_score") != 0)
		err(EXIT_FAILURE, _("failed to read OOM score value"));

	return ret;
}

static int get_score_adj(struct path_cxt *pc)
{
	int ret;

	if (ul_path_read_s32(pc, &ret, "oom_score_adj") != 0)
		err(EXIT_FAILURE, _("failed to read OOM score adjust value"));

	return ret;
}

static int set_score_adj(struct path_cxt *pc, int adj)
{
	return ul_path_write_s64(pc, adj, "oom_score_adj");
}

int main(int argc, char **argv)
{
	pid_t pid = 0;
	int c, adj = 0, has_adj = 0;
	struct path_cxt *pc = NULL;

	static const struct option longopts[] = {
		{ "adjust",  required_argument, NULL, 'n' },
		{ "pid",     required_argument, NULL, 'p' },
		{ "help",    no_argument,       NULL, 'h' },
		{ "version", no_argument,       NULL, 'V' },
		{ NULL,      0,                 NULL,  0  }
	};

	setlocale(LC_ALL, "");
	bindtextdomain(PACKAGE, LOCALEDIR);
	textdomain(PACKAGE);
	close_stdout_atexit();

	while ((c = getopt_long(argc, argv, "hn:p:V", longopts, NULL)) != -1) {
		switch (c) {
		case 'p':
			pid = strtos32_or_err(optarg, _("invalid PID argument"));
			break;
		case 'n':
			adj = strtos32_or_err(optarg, _("invalid adjust argument"));
			has_adj = 1;
			break;

		case 'V':
			print_version(EXIT_SUCCESS);
		case 'h':
			usage();
		default:
			errtryhelp(EXIT_FAILURE);
		}
	}

	if (optind < argc && pid) {
		warnx(_("invalid argument: %s"), argv[optind]);
		errtryhelp(EXIT_FAILURE);
	}
	if (!pid && argc - optind < 1) {
		warnx(_("no PID or COMMAND specified"));
		errtryhelp(EXIT_FAILURE);
	}
	if (optind < argc && !has_adj) {
		warnx(_("no OOM score adjust value specified"));
		errtryhelp(EXIT_FAILURE);
	}

	pc = ul_new_path("/proc/%d", (int) (pid ? pid : getpid()));

	/* Show */
	if (!has_adj) {
		printf(_("pid %d's current OOM score: %d\n"), pid, get_score(pc));
		printf(_("pid %d's current OOM score adjust value: %d\n"), pid, get_score_adj(pc));

	/* Change */
	} else if (pid) {
		int old = get_score_adj(pc);

		if (set_score_adj(pc, adj))
			err(EXIT_FAILURE, _("failed to set score adjust value"));

		printf(_("pid %d's OOM score adjust value changed from %d to %d\n"), pid, old, adj);

	/* Start new process */
	} else {
		if (set_score_adj(pc, adj))
			err(EXIT_FAILURE, _("failed to set score adjust value"));
		ul_unref_path(pc);
		argv += optind;
		execvp(argv[0], argv);
		errexec(argv[0]);
	}

	ul_unref_path(pc);
	return EXIT_SUCCESS;
}
